numeric_ranges_algorithms/numeric_ranges_algorithms.md

```diff
@@ -0,0 +1,484 @@
+
+### We don't propose `reduce_with_iter`
+
+A `reduce_with_iter` algorithm would look like `fold_left_with_iter`, but would permit reordering of binary operations.  It would return both an iterator to one past the last element, and the computed value.  A hypothetical `reduce_with_iter` algorithm would also return an iterator to one past the last element, and the computed value, but would share `reduce`'s permission to reorder binary operations.
+
+We do not propose the analogous `reduce_with_iter` here, though we would not oppose someone else proposing it.  That algorithm would serve users who are writing code generic enough to work with single-pass input iterators, _and_ who want to expose potential binary operation reordering opportunities.
+
+Just like `fold_left`, the `reduce` algorithm should return just the computed value.  Section 4.4 of <a href="https://wg21.link/P2322R6">P2322R6</a> argues that this makes it easier to use, and improves consistency with other ranges algorithms like `ranges::count` and `ranges::any_of`.  It is also consistent with P3179R8.  The algorithms `fold_left_with_iter` and `fold_left_first_with_iter` exist for users who want both the iterator and the value.  Section 4.4 of P2322R6 further elaborates that `fold_left` should not be specified in terms of `fold_left_with_iter`, for performance reasons: it would "incur an extra move of the accumulated result, due to lack of copy elision (we have different return types)."  The `*_with_iter` algorithms are separate algorithms that need separate specifications.
+
+### Algorithms that we propose here
+
+This leaves three algorithms, which we propose here: `reduce`, `inclusive_scan`, and `exclusive_scan`.  <a href="https://wg21.link/P2760R1">P2760R1</a> proposes convenience wrappers `ranges::sum(r)` as `ranges::reduce(r, plus{}, range_value_t<R>())` and `ranges::product(r)` as `ranges::reduce(r, multiplies{}, range_value_t<R>(1))`; we propose parallel and non-parallel overloads of these here as well.
+
+## Range categories and return types
+
+We make the parallel algorithms proposed here take sized random access ranges.  As a result, any parallel algorithms with an output range need to return an iterator to one past the last element of the input, in case there wasn't enough room in the output.  We make the non-parallel algorithms take (single-pass) sized `input_range` and `output_range`, for consistency with the C++17 parallel algorithms that only require _LegacyInputIterator_ and _LegacyOutputIterator_.  This suggests making the algorithms with output ranges -- `inclusive_scan` and `exclusive_scan` -- return both input and output iterators.  This is consistent both with the existing non-parallel ranges algorithms and with P3179R8 (e.g., `ranges::transform`).

```

Comment:
> We make the non-parallel algorithms take (single-pass) sized `input_range` and `output_range`, for consistency with the C++17 parallel algorithms that only require _LegacyInputIterator_ and _LegacyOutputIterator_.

C++17 parallel algorithms in `<numeric>` require at least `ForwardIterator`. Did you mean the non-parallel algorithms?

Reply:
That's right -- thanks for the correction!